Add cross-platform NTLM proxy authentication support

nollium opened the pull request:
# Add cross-platform NTLM proxy authentication support

Previously, Windows authentication for corporate proxies was limited to Windows-based payloads only. This PR extends NTLM proxy authentication support to all operating systems, including Linux.

The implementation adds a new `--ntlm-proxy-creds` flag to both the server's `link` command and the `reverse_ssh` generated payload.

## Usage

Server-side linking:
```sh
link --proxy http://proxy.corporate.local:8080 --ntlm-proxy-creds 'DOMAIN\USER:PASS'
```

Payload execution:
```sh
./rssh-payload --proxy http://proxy.corporate.local:8080 --ntlm-proxy-creds 'DOMAIN\USER:PASS'
```

The NTLM implementation is taken from https://github.com/bodgit/ntlmssp

I've tested this new feature in labs and real corporate environments and didn't encounter any issues for now. 

I've also rewritten the server's command-line parsing so it properly handles single quotes and backslashes like a real shell would do, because the old parsing caused problems with the domain credentials syntax `DOMAIN\USER:PASS`, backslashes had to be double escaped, no matter what kind of quotes were used, which would be misleading for users.

This is my first open-source contribution, and I welcome any feedback to improve the implementation.

NHAS commented:
Awesome! I had a look at your code a little while ago and wanted to give it a merge but ran into a few things I had to do otherwise. 

I think this is real nice! 

I have a few questions:
- Why have you put the ntlmssp library code into rssh? Have you made changes to it?
- I dont suppose there are any tests you could add to the e2e tests to make sure this functionality keeps working in to the future? (Dont worry if not!)

NHAS commented:
Just adding another ping to this. The only thing stopping me for merging this in is just the question around the ntlmssp library being added wholesale rather than just imported as a module

nollium commented:
Hey! Thanks for the feedback. I've removed the ntlmssp code from the codebase and now properly import it from GitHub as a dependency.

I've added some e2e tests for the NTLM proxy auth - though I want to be upfront that while I can work with auth systems, I'm not really an expert on NTLM protocol specifics. I put together the test cases by referencing other implementations and docs, and while everything seems to work correctly both in testing and real corporate environments, I'd definitely welcome any expert eyes on the implementation.

NHAS commented:
Well dang this is absolutely fantastic, Im going to give this a merge into the unstable branch and give it a test myself. Then I'll get it in main post haste

NHAS commented:
Can you please explain the line:

```go
challengeStr := string(responseStatus[challengeStart+5:])
```

As it has no guard or explanation, and may cause a crash if the proxy is bad

NHAS commented:
Im assuming its for the "NTLM " string

NHAS commented:
I've made some little cleanups on the unstable branch. Feel free to test them and once happy I'll give it a merge
